server/schedule/placement/rule_manager.go

```diff
@@ -437,5 +442,148 @@ func (m *RuleManager) DeleteRuleGroup(id string) error {
 	defer m.Unlock()
 	p := m.ruleConfig.beginPatch()
 	p.deleteGroup(id)
-	return m.tryCommitPatch(p)
+	if err := m.tryCommitPatch(p); err != nil {
+		return err
+	}
+	log.Info("group config reset", zap.String("group", id))
+	return nil
+}
+
+// GetAllGroupBundles returns all rules and groups configuration. Rules are
+// grouped by groups.
+func (m *RuleManager) GetAllGroupBundles() []GroupBundle {
+	m.RLock()
+	defer m.RUnlock()
+	var bundles []GroupBundle
+	for _, g := range m.ruleConfig.groups {
+		bundles = append(bundles, GroupBundle{
+			ID:       g.ID,
+			Index:    g.Index,
+			Override: g.Override,
+		})
+	}
+	for _, r := range m.ruleConfig.rules {
+		for i := range bundles {
+			if bundles[i].ID == r.GroupID {
+				bundles[i].Rules = append(bundles[i].Rules, r)
+			}
+		}
+	}
+	sort.Slice(bundles, func(i, j int) bool {
+		return bundles[i].Index < bundles[j].Index ||
+			(bundles[i].Index == bundles[j].Index && bundles[i].ID < bundles[j].ID)
+	})
+	for _, b := range bundles {
+		sortRules(b.Rules)
+	}
+	return bundles
+}
+
+// GetGroupBundle returns a group and all rules belong to it.
+func (m *RuleManager) GetGroupBundle(id string) (b GroupBundle) {
+	m.RLock()
+	defer m.RUnlock()
+	b.ID = id
+	if g := m.ruleConfig.groups[id]; g != nil {
+		b.Index, b.Override = g.Index, g.Override
+	}
+	for _, r := range m.ruleConfig.rules {
+		if r.GroupID == id {
+			b.Rules = append(b.Rules, r)
+		}
+	}
+	sortRules(b.Rules)
+	return
+}
+
+// SetAllGroupBundles resets full configuration. All old configurations are dropped.
+func (m *RuleManager) SetAllGroupBundles(groups []GroupBundle) error {
+	m.Lock()
+	defer m.Unlock()
+	p := m.ruleConfig.beginPatch()
+	for k := range m.ruleConfig.rules {
+		p.deleteRule(k[0], k[1])
+	}
```

Comment:
I think there can be an api to delete all old rules/groups for ruleConfig. Why iterate over all rules, if it can be done by reslicing.

Reply:
We cannot simply clear the map because we need to record all delete operations in the patch and later we need to apply the operations to persistent storage.

Reply:
Oh, yes. Persistence needs the records.